agent/consul/catalog_endpoint.go

```diff
@@ -672,10 +672,34 @@ func (c *Catalog) ServiceNodes(args *structs.ServiceSpecificRequest, reply *stru
 				return err
 			}
 
-			reply.Index, reply.ServiceNodes = index, services
+			mergedServices := services
+
+			if args.MergeCentralConfig {
+				var mergedServiceNodes structs.ServiceNodes
+				for _, sn := range services {
+					mergedsn := sn
+					ns := sn.ToNodeService()
+					if ns.IsSidecarProxy() || ns.IsGateway() {
```

Comment:
Calling out that the merging would happen only for mesh services to mirror when the agent registration performs merging [here](https://github.com/hashicorp/consul/blob/main/agent/agent.go#L2072-L2073).